lsif/src/worker/conversion/importer.ts

```diff
@@ -437,6 +445,78 @@ function getReferences(correlator: Correlator): SymbolReferences[] {
     }))
 }
 
+/**
+ * Merge the data in the correlator of all documents that share the same path. This
+ * function works by moving the contains, definition, and reference data keyed by a
+ * document with a duplicate path into a canonical document with that path. The first
+ * document inserted for a path is the canonical document for that path. This function
+ * guarantees that duplicate document ids are removed from these maps.
+ *
+ * @param correlator The correlator with all vertices and edges inserted.
+ */
+function mergeDocuments(correlator: Correlator): void {
+    const uriMap = new Map<string, sqliteModels.DocumentId>()
+    for (const [id, path] of correlator.documentPaths.entries()) {
+        const canonicalId = uriMap.get(path)
+        if (canonicalId === undefined) {
+            uriMap.set(path, id)
+            continue
+        }
+
+        mergeContains(id, canonicalId, correlator.containsData)
+        mergeDefinitionReferences(id, canonicalId, correlator.definitionData)
+        mergeDefinitionReferences(id, canonicalId, correlator.referenceData)
+        correlator.documentPaths.delete(id)
+    }
+}
+
+/**
+ * Move the contains data for document `id` into the contains data of document
+ * `canonicalId`, then delete the reference to document `id.`
+ *
+ * @param id The id of the document to replace.
+ * @param canonicalId The id of the document to subsume the other.
+ * @param containsData The containment map of the correlator.
+ */
+function mergeContains(
+    id: sqliteModels.DocumentId,
+    canonicalId: sqliteModels.DocumentId,
+    containsData: Map<sqliteModels.DocumentId, Set<lsif.RangeId>>
+): void {
+    const contains = mustGet(containsData, id, 'contains')
+    const canonicalContains = mustGet(containsData, canonicalId, 'contains')
+    for (const rangeId of contains) {
+        canonicalContains.add(rangeId)
+    }
+    containsData.delete(id)
```

Comment:
Just to save space? (That's fine if so, just checking if there's a correctness concern we should document)

Reply:
I think the only deletion that's necessary is the document data, which makes it so we don't insert one of the documents subsumed by the canonical representative. The ones for the contains and def/ref results may be unnecessary.

Reply:
Makes sense. It's probably worth a comment to say so.

```suggestion
    // Discard the document data as a flag to prevent inserting one of the
    // documents subsumed by the canonical representative.
    containsData.delete(id)
```

Reply:
Added comments (but this one was attached to the wrong delete :p)

Reply:
Ah, yes—sorry, I tunnel-visioned on the place where I was writing _my_ comment. Thanks!